File: sdk-python/examples/json/json_example.py
import asyncio
import logging
from pathlib import Path
import random
from typing import Any
import littlehorse

from littlehorse.config import LHConfig
from littlehorse.model.common_enums_pb2 import VariableType
from littlehorse.worker import LHTaskWorker, WorkerContext
from littlehorse.workflow import ThreadBuilder, Workflow

logging.basicConfig(level=logging.INFO)


def get_config() -> LHConfig:
    config = LHConfig()
    config_path = Path.home().joinpath(".config", "littlehorse.config")
    if config_path.exists():
        config.load(config_path)
    return config


def get_workflow() -> Workflow:
    def my_entrypoint(thread: ThreadBuilder) -> None:
        person = thread.add_variable("person", VariableType.JSON_OBJ)
        thread.execute("greet", person.with_json_path("$.name"))
        thread.execute("describe-car", person.with_json_path("$.car"))

    return Workflow("example-json", my_entrypoint)


async def greeting(name: str, ctx: WorkerContext) -> str:
    msg = f"Hello {name}!. WfRun {ctx.wf_run_id}"
    print(msg)
    await asyncio.sleep(random.uniform(0.5, 1.5))
    return msg


async def describe_car(car: dict[str, Any]) -> str:
    msg = f"You drive a {car['brand']} model {car['model']}"
    print(msg)
    await asyncio.sleep(random.uniform(0.5, 1.5))
    return msg


async def main() -> None:
    config = get_config()
    wf = get_workflow()

    littlehorse.register_task(greeting, "greet", config)
    littlehorse.register_task(describe_car, "describe-car", config)
    littlehorse.register_workflow(wf, config)

    await littlehorse.start(
        LHTaskWorker(greeting, "greet", config),
        LHTaskWorker(describe_car, "describe-car", config),
    )


if __name__ == "__main__":
    asyncio.run(main())

File: examples/java/json/src/main/java/io/littlehorse/examples/JsonExample.java
package io.littlehorse.examples;

import io.littlehorse.sdk.common.config.LHConfig;
import io.littlehorse.sdk.common.proto.LittleHorseGrpc;
import io.littlehorse.sdk.common.proto.VariableType;
import io.littlehorse.sdk.wfsdk.WfRunVariable;
import io.littlehorse.sdk.wfsdk.Workflow;
import io.littlehorse.sdk.wfsdk.internal.WorkflowImpl;
import io.littlehorse.sdk.worker.LHTaskWorker;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.Properties;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/*
 * This workflow demonstrates the ability of the Java SDK to serialize/deserialize
 * JSON and allow Task developers to interact with real Java objects.
 * More information about json-path at https://github.com/json-path/JsonPath.
 *
 * To run this workflow:
 * `lhctl run json-example person '{"name": "Obi-Wan", "car": {"brand": "Ford", "model": "Escape"}}'`
 */
public class JsonExample {

    private static final Logger log = LoggerFactory.getLogger(JsonExample.class);

    public static Workflow getWorkflow() {
        return new WorkflowImpl("example-json", wf -> {
            WfRunVariable person = wf.declareJsonObj("person");

            wf.execute("greet", person.jsonPath("$.name"));
            wf.execute("describe-car", person.jsonPath("$.car"));
        });
    }

    public static Properties getConfigProps() throws IOException {
        Properties props = new Properties();
        File configPath = Path.of(System.getProperty("user.home"), ".config/littlehorse.config")
                .toFile();
        if (configPath.exists()) {
            props.load(new FileInputStream(configPath));
        }
        return props;
    }

    public static List<LHTaskWorker> getTaskWorkers(LHConfig config) {
        CarTaskWorker executable = new CarTaskWorker();
        List<LHTaskWorker> workers = List.of(
                new LHTaskWorker(executable, "greet", config), new LHTaskWorker(executable, "describe-car", config));

        // Gracefully shutdown
        Runtime.getRuntime()
                .addShutdownHook(new Thread(() -> workers.forEach(worker -> {
                    log.debug("Closing {}", worker.getTaskDefName());
                    worker.close();
                })));
        return workers;
    }

    public static void main(String[] args) throws IOException {
        // Let's prepare the configurations
        Properties props = getConfigProps();
        LHConfig config = new LHConfig();
        LittleHorseGrpc.LittleHorseBlockingStub client = config.getBlockingStub();

        // New workflow
        Workflow workflow = getWorkflow();

        // New workers
        List<LHTaskWorker> workers = getTaskWorkers(config);

        // Register tasks
        for (LHTaskWorker worker : workers) {
            worker.registerTaskDef();
        }

        // Register a workflow
        workflow.registerWfSpec(client);

        // Run the workers
        for (LHTaskWorker worker : workers) {
            log.debug("Starting {}", worker.getTaskDefName());
            worker.start();
        }
    }
}
